File: algorithms/array/product_of_array_except_self.py
# https://leetcode.com/problems/product-of-array-except-self/

# time complexity: O(n)

class Solution:
    def productExceptSelf(self, nums: list[int]) -> list[int]:
        multiplied = []
        m = 1
        for n in nums:
            multiplied.append(m)
            m *= n
        m = 1
        for i in range(len(nums) - 1, -1, -1):
            multiplied[i] *= m
            m *= nums[i]
        return multiplied

File: leetcode/array/product_of_array_except_self.cc
// https://leetcode.com/problems/product-of-array-except-self/

// time complexity: O(n)

#include <vector>

using namespace std;

class Solution {
public:
    vector<int> productExceptSelf(vector<int> &nums) {
        int len = nums.size();
        vector<int> multiplied(len);
        for (int i = 0, m = 1; i < len; i++) {
            multiplied[i] = m;
            m *= nums[i];
        }
        for (int i = len - 1, m = 1; i >= 0; i--) {
            multiplied[i] *= m;
            m *= nums[i];
        }
        return multiplied;
    }
};